### DISCUSSION TITLE
Questions about mpi_reduce of GROOPS code

### DISCUSSION THREAD

**Discussion by FFFFFG-IGG** [groops-devs/groops#178] [Q&A] [answered]
Hello,
  I've lost in understanding mpi reduce of groops code in gnss.cpp:
![image](https://user-images.githubusercontent.com/59271400/231638712-aa51fbe4-e80e-44f2-baf7-8a5b9ac8d3fc.png)
here, each rank will totally finish its loop, which means if I deploy 4 process, Vector 'recvProcess' of its rank will be filled with its rank number(code here is rank+1:1,2,3,4). Then reduce will sum all Vector from the 'start pointer' till the 'end pointer', and get a recvProcess filled with '10', just as what the function 'reduce' defined: 
![image](https://user-images.githubusercontent.com/59271400/231639637-2c1423e6-1616-4e15-ae49-8b1085bcaf14.png)
![image](https://user-images.githubusercontent.com/59271400/231639728-1e406332-5898-42fd-b4e4-55825ea0f7db.png)

 However elements of recvProcess shouldn't be '10' but stachastic '1','2','3' or 4' according the following loop conditions.
 here comes the question: 
1.recvProcess elements are 10 or the other circumstance?
2.If recvProcess elements are stochastic '1','2','3' or 4' , how does groops do this?

I'll appreciate if someone could offer help.

**Comment by tmayerguerr** [comment, marked as answer]
Hi,
The receivers are distributed to the nodes. For example, receiver 0 on node 0, receiver 1 on node 1, and so on. However, the processes do not know on which node the other receivers are. This information is distributed with recvProcess.
In the first loop, the number is only set on the node of the receiver (`if(recv->isMyRank())`), on all others the value remains 0:
```
process0: recvProcess[1,0,0,0...]
process1: recvProcess[0,2,0,0...]
process2: recvProcess[0,0,3,0...]
process3: recvProcess[0,0,0,4...]
```
So the result after reduceSum and broadCast is
```
process0: recvProcess[1,2,3,4...]
process1: recvProcess[1,2,3,4...]
process2: recvProcess[1,2,3,4...]
process3: recvProcess[1,2,3,4...]
```
Unused/disabled receivers have the value 0. 
Hope this helps.

regards
Torsten

**Reply by FFFFFG-IGG** [reply]
> Hi, The receivers are distributed to the nodes. For example, receiver 0 on node 0, receiver 1 on node 1, and so on. However, the processes do not know on which node the other receivers are. This information is distributed with recvProcess. In the first loop, the number is only set on the node of the receiver (`if(recv->isMyRank())`), on all others the value remains 0:
> 
> ```
> process0: recvProcess[1,0,0,0...]
> process1: recvProcess[0,2,0,0...]
> process2: recvProcess[0,0,3,0...]
> process3: recvProcess[0,0,0,4...]
> ```
> 
> So the result after reduceSum and broadCast is
> 
> ```
> process0: recvProcess[1,2,3,4...]
> process1: recvProcess[1,2,3,4...]
> process2: recvProcess[1,2,3,4...]
> process3: recvProcess[1,2,3,4...]
> ```
> 
> Unused/disabled receivers have the value 0. Hope this helps.
> 
> regards Torsten

Oh, I got it! I missunderstood that receivers->isMyrank() is always true after the receivers init. Many thanks！
